### DISCUSSION TITLE
permissions to create schema from the client

### DISCUSSION THREAD

**Discussion by edwardLz** [supabase/supabase#957] [Questions] [answered]
Is there any way to create a schema, with a stored procedure? I am currently trying but it tells me that my database has no permissions.

**Comment by silentworks** [comment, marked as answer]
This is a weird one, schemas are normally not dynamically created, they are thought out in advance and created. Can you state your use case? then maybe a better solution can be provided.

**Reply by edwardLz** [reply]
In my platform I need to create organizations that will have access to it, for each new organization, I need to create a scheme where a lot of information will be saved, this scheme already has its table and structure "thought", that is, the scheme is already done, it only repeats each organization.

That is why I need to create diagrams dynamically, eye, the structure of those diagrams is already thought, only the same structure is repeated for each diagram, the thing is that there is a lot of information that needs to be saved, that is why diagrams are created by organization .

If you have another idea, I would really appreciate it. :+1:

**Reply by steve-chavez** [reply]
@edwardLz Do you mean you have schema-based multitenancy in your application?

(Check this [article](https://web.archive.org/web/20080510114523/https://msdn.microsoft.com/en-us/library/aa479086.aspx) for knowing the types of multitenancy on a db)

If that's the case, an alternative could also be a shared schema for all the tenants. Here you could use RLS to make sure a tenant cannot see other's tenants rows.

---

That being said, it should be possible to create a schema from within a stored procedure. You could create that stored procedure  as [SECURITY DEFINER](https://www.postgresql.org/docs/current/sql-createfunction.html#SQL-CREATEFUNCTION-SECURITY) and make sure only privileged users have access to that function(`GRANT EXECUTE`)

**Reply by silentworks** [reply]
I would suggest the same as @steve-chavez in using a shared schema with RLS to keep them secured. Also another option would be to use the `json` column type for certain data. It's a bit hard to recommend clearly as we don't know your full requirements.

**Reply by edwardLz** [reply]
Ok I understand. It is very clear to me, the article helped me a lot.

.Thank you so much guys.

**Reply by edwardLz** [reply]
Hello again. just to be sure ... in order to access these tables by running queries from my application, would I have to do it through stored procedures?
